src/Microsoft.OData.Core/WriterValidationUtils.cs

```diff
@@ -314,6 +316,83 @@ internal static void ValidateStreamReferenceProperty(ODataProperty streamPropert
             }
         }
 
+        /// <summary>
+        /// Validates the value type of a property meets the derived type constraints.
+        /// </summary>
+        /// <param name="propertySerializationInfo">The property serialization info.</param>
+        /// <remarks>This does NOT validate the value of the property, just the type of property.</remarks>
+        internal static void ValidatePropertyDerivedTypeConstraint(PropertySerializationInfo propertySerializationInfo)
+        {
+            Debug.Assert(propertySerializationInfo != null, "propertySerializationInfo != null");
+
+            // Skip the undeclared property
+            if (propertySerializationInfo.MetadataType.IsUndeclaredProperty)
+            {
+                return;
+            }
+
+            PropertyValueTypeInfo valueType = propertySerializationInfo.ValueType;
+            if (valueType == null || valueType.TypeReference == null)
+            {
+                return;
+            }
+
+            // make sure the same type can pass the validation.
+            if (propertySerializationInfo.MetadataType.TypeReference.Definition == valueType.TypeReference.Definition)
+            {
+                return;
+            }
+
+            string fullTypeName = valueType.TypeReference.FullName();
+            if (propertySerializationInfo.MetadataType.DerivedTypeConstraints == null ||
+                propertySerializationInfo.MetadataType.DerivedTypeConstraints.Any(d => d == fullTypeName))
+            {
+                return;
+            }
+
+            throw new ODataException(Strings.WriterValidationUtils_ValueTypeNotAllowedInDerivedTypeConstraint(fullTypeName, "property", propertySerializationInfo.PropertyName));
+        }
+
+        /// <summary>
+        /// Validates the value type of a property meets the derived type constraints.
+        /// </summary>
+        /// <param name="propertySerializationInfo">The property serialization info.</param>
+        /// <remarks>This does NOT validate the value of the property, just the type of property.</remarks>
+        internal static void ValidateCollectionPropertyDerivedTypeConstraint(ODataCollectionValue collectionValue,

```

Comment:
>ValidateCollectionPropertyDerivedTypeConstraint [](start = 29, length = 47)

Can this be combined with ValidatePropertyDerivedTypeConstraint? It looks like the logic is exactly the same, and that the ODataCollectionValue is not used.

Reply:
#Resolved. That's old codes to verify the collection(Edm.PrimitiveType). Not, it's not valid. And I removed. Thanks.